util/design/lib/OtpMemImg.py

```diff
@@ -0,0 +1,384 @@
+        raise RuntimeError('Key is out of 128bit range')
+
+    # Make sure inputs are integers
+    if not isinstance(plain, int) or not isinstance(key, int):
+        raise RuntimeError('Data and key need to be of type int')
+
+    cipher = Present(_int_to_byte_array(key, 16))
+    encr = cipher.encrypt(_int_to_byte_array(plain, 8))
+    return _byte_array_to_int(encr)
+
+
+def _present_64bit_digest(data_blocks, iv, const):
+    '''Compute digest over multiple 64bit data blocks'''
+
+    # We need to align the number of data blocks to 2x64bit
+    # for the digest to work properly.
+    if len(data_blocks) % 2 == 1:
+        data_blocks.append(data_blocks[-1])
+
+    # Merge neighboring 64bit blocks into 128bit block
+    data_blocks_merged = []
+    for k, block in enumerate(data_blocks):
+        if (k % 2) == 0:
+            data_blocks_merged.append(block)
+        else:
+            data_blocks_merged[-1] += (block << 64)
+
+    # This computes a digest according to a Merkle-Damgard construction
+    # that uses the Davies-Meyer scheme to turn the PRESENT cipher into
+    # a one-way compression function. Digest finalization consists of
+    # a final digest round with a 128bit constant.
+    # See also: https://docs.opentitan.org/hw/ip/otp_ctrl/doc/index.html#scrambling-datapath
+    state = iv
+    for block in data_blocks_merged:
+        state ^= _present_64bit_encrypt(state, block)
+
+    return state
+
+
+def _to_hexfile_with_ecc(data, annotation, config):
+    '''Compute ECC and convert into memory hexfile'''
+
+    log.info('Convert to HEX file.')
+
+    if config['secded']['data_width'] > 64:
+        raise RuntimeError('OTP data width cannot be larger than 64')
+
+    num_words = len(data) * 8 // config['secded']['data_width']
+    bytes_per_word = config['secded']['data_width'] // 8
+    # Byte aligned total width after adding ECC bits
+    bytes_per_word_ecc = (
+        (config['secded']['data_width'] + config['secded']['ecc_width'] + 7) //
+        8)
+    bin_format_str = '0' + str(config['secded']['data_width']) + 'b'
+    hex_format_str = '0' + str(bytes_per_word_ecc * 2) + 'x'
+    memory_words = '// OTP memory hexfile with {} x {}bit layout\n'.format(
+        num_words, bytes_per_word_ecc * 8)
+    log.info('Memory layout is {} x {}bit (with ECC)'.format(
+        num_words, bytes_per_word_ecc * 8))
+
+    for k in range(num_words):
+        # Assemble native OTP word and uniquify annotation for comments
+        word = 0
+        word_ann = OrderedDict()
+        for j in range(bytes_per_word):
+            idx = k * bytes_per_word + j
+            word += data[idx] << (j * 8)
+            if annotation[idx] not in word_ann:
+                word_ann.update({annotation[idx]: "1"})
+
+        # This prints the byte offset of the corresponding
+        # payload data in the memory map (excluding ECC bits)
+        word_ann = ' // {:06x}: '.format(k * bytes_per_word) + ', '.join(
+            word_ann.keys())
```

Comment:
Maybe rename this variable? At the moment, it's a string that replaces a dictionary, which might be a bit confusing.

Reply:
yeah makes sense - done!